xmodule/services.py

```diff
@@ -308,3 +312,147 @@ def _handle_deprecated_progress_event(self, block, event):
         # in order to avoid duplicate work and possibly conflicting semantics.
         if not getattr(block, 'has_custom_completion', False):
             self.completion_service.submit_completion(block.scope_ids.usage_id, 1.0)
+
+
+class ProblemFeedbackService(Service):
+    """
+    An XBlock Service that allows XModules to define:
+
+    * If the answers are visible.
+    * Whether correctness is currently hidden. When correctness is hidden, this
+      limits the user's access to the correct/incorrect flags, messages, problem
+      scores, and aggregate subsection and course grades.
+    """
+    def __init__(self, block=None, user_is_staff=False, **kwargs):
+        super().__init__(**kwargs)
+        self._user_is_staff = user_is_staff
+        # This is needed because the `Service` class initialization expects the XBlock passed as an `xblock` keyword
+        #  argument, but the `service` method from the `DescriptorSystem` passes a `block`.
+        self._xblock = self.xblock() or block
+
+    def is_past_due(self):
+        """
+        Returns if the problem is past its due date.
+
+        This method returns True when an XBlock doesn't have a due date. That
+        is because if an XBlock doesn't have a due date, and we are trying to
+        assess whether we can display a correct answer to a learner, we want to
+        treat an answered problem without a due date as past due. Otherwise, we
+        can never show the answer to a learner, since there is no due date.
+
+        There are scenarios where you would want is_past_due to return False
+        when an XBlock doesn't have a due date, for example, when determining
+        whether a learner should still be able to submit their answer.
+        """
+        if not self._xblock:
+            return False
+
+        due_date = getattr(self._xblock, 'close_date', None) or getattr(self._xblock, 'due', None)
+        return (due_date is None or datetime.now(UTC) > due_date)
+
+    def is_attempted(self):
+        """
+        Has the problem been attempted?
+        """
+        if not self._xblock:
+            return False
+
+        return self._xblock.attempts > 0
+
+    def used_all_attempts(self):
+        """ All attempts have been used """
+        if not self._xblock:
+            return False
+
+        return (self._xblock.max_attempts is not None and
+                self._xblock.attempts >= self._xblock.max_attempts)
+
+    def closed(self):
+        """
+        Is the student still allowed to submit answers?
+        """
+        if not self._xblock:
+            return False
+
+        if self.used_all_attempts():
+            return True
+        # Using XBlock's is_past_due method, as it's suited for checking the
+        # submission deadline, unlike the is_past_due method of this service.
+        elif self._xblock.is_past_due():
+            return True
```

Comment:
Where does the discrepancy between `ProblemFeedbackService.is_past_due` and `InheritableFieldsMixin.is_past_due` come from? Having two methods with the same name that behave differently is confusing.